File: src/main/java/com/skysoft/gui/scale/GuiScaleController.java
package com.skysoft.gui.scale;

import com.mojang.blaze3d.platform.Window;
import com.skysoft.config.InventoryScreenConfig;
import com.skysoft.config.SkysoftConfigGui;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;
import net.minecraft.client.renderer.state.gui.GuiRenderState;

public final class GuiScaleController {
    private static final Minecraft MINECRAFT = Minecraft.getInstance();

    private static RenderBatch pendingRenderBatch;
    private static boolean overlaysUseNormalCoordinates;

    private GuiScaleController() {
    }

    public static boolean scalesInventory(Screen screen) {
        return config().separateInventoryGuiScale && supportsInventoryScale(screen);
    }

    public static boolean scalesTooltip(Screen screen) {
        return config().separateTooltipGuiScale && supportsInventoryScale(screen);
    }

    public static ResolvedScales resolve(Screen screen, Window window) {
        int normal = resolve(window, MINECRAFT.options.guiScale().get());
        int tooltip = resolve(window, config().settings.tooltipGuiScale);
        int inventory = Math.min(
            resolve(window, config().settings.inventoryGuiScale),
            inventoryScaleLimit(screen)
        );
        return new ResolvedScales(normal, Math.max(1, inventory), tooltip);
    }

    public static WindowScaleOverride useInventoryScale(Screen screen, Window window) {
        return new WindowScaleOverride(window, resolve(screen, window).inventory());
    }

    public static WindowScaleOverride useTooltipScale(Screen screen, Window window) {
        return new WindowScaleOverride(window, resolve(screen, window).tooltip());
    }

    public static int convertCoordinate(int coordinate, int sourceScale, int targetScale) {
        return Math.round(coordinate * Math.max(1, sourceScale) / (float) Math.max(1, targetScale));
    }

    public static void updateScreenDimensions(Screen screen, Window window) {
        ScaledScreenState state = (ScaledScreenState) screen;
        int width = window.getGuiScaledWidth();
        int height = window.getGuiScaledHeight();
        if (!state.skysoft$matchesScaleDimensions(width, height)) {
            screen.resize(width, height);
            state.skysoft$rememberScaleDimensions(width, height);
        }
    }

    public static void restoreScreenDimensions(Screen screen, Window window) {
        ScaledScreenState state = (ScaledScreenState) screen;
        if (!state.skysoft$hasScaleDimensions()) {
            return;
        }

        int width = window.getGuiScaledWidth();
        int height = window.getGuiScaledHeight();
        if (!state.skysoft$matchesScaleDimensions(width, height)) {
            screen.resize(width, height);
        }
        state.skysoft$forgetScaleDimensions();
    }

    public static void submitRenderBatch(GuiRenderState inventory, GuiRenderState overlays) {
        if (pendingRenderBatch != null) {
            throw new IllegalStateException("Inventory GUI render batch was not consumed");
        }
        pendingRenderBatch = new RenderBatch(inventory, overlays);
    }

    public static RenderBatch takeRenderBatch() {
        RenderBatch batch = pendingRenderBatch;
        pendingRenderBatch = null;
        return batch;
    }

    public static boolean overlaysUseNormalCoordinates() {
        return overlaysUseNormalCoordinates;
    }

    public static void setOverlaysUseNormalCoordinates(boolean value) {
        overlaysUseNormalCoordinates = value;
    }

    private static int resolve(Window window, int configuredScale) {
        return Math.max(1, window.calculateScale(Math.max(0, configuredScale), MINECRAFT.isEnforceUnicode()));
    }

    private static boolean supportsInventoryScale(Screen screen) {
        return screen instanceof AbstractContainerScreen<?> ||
            screen instanceof InventoryScaledScreen scaledScreen && scaledScreen.usesInventoryScale();
    }

    private static int inventoryScaleLimit(Screen screen) {
        if (screen instanceof InventoryScaledScreen scaledScreen && scaledScreen.usesInventoryScale()) {
            return Math.max(1, scaledScreen.inventoryScaleLimit());
        }
        return Integer.MAX_VALUE;
    }

    private static InventoryScreenConfig config() {
        return SkysoftConfigGui.INSTANCE.config().gui.inventoryScreen;
    }

    public record ResolvedScales(int normal, int inventory, int tooltip) {
    }

    public record RenderBatch(GuiRenderState inventory, GuiRenderState overlays) {
    }

    public static final class WindowScaleOverride implements AutoCloseable {
        private final Window window;
        private final int previousScale;
        private boolean closed;

        private WindowScaleOverride(Window window, int requestedScale) {
            this.window = window;
            previousScale = window.getGuiScale();
            window.setGuiScale(requestedScale);
        }

        @Override
        public void close() {
            if (closed) {
                throw new IllegalStateException("GUI scale override closed twice");
            }
            closed = true;
            window.setGuiScale(previousScale);
        }
    }
}

File: src/main/kotlin/com/skysoft/gui/scale/GuiScaleController.kt
package com.skysoft.gui.scale

import com.mojang.blaze3d.platform.Window
import com.skysoft.config.InventoryScreenConfig
import com.skysoft.config.SkysoftConfigGui
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.screens.Screen
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen
import net.minecraft.client.renderer.state.gui.GuiRenderState

class GuiScaleController private constructor() {
    data class ResolvedScales(
        private val normalValue: Int,
        private val inventoryValue: Int,
        private val tooltipValue: Int,
    ) {
        fun normal(): Int = normalValue

        fun inventory(): Int = inventoryValue

        fun tooltip(): Int = tooltipValue
    }

    data class RenderBatch(
        private val inventoryValue: GuiRenderState,
        private val overlaysValue: GuiRenderState,
    ) {
        fun inventory(): GuiRenderState = inventoryValue

        fun overlays(): GuiRenderState = overlaysValue
    }

    class WindowScaleOverride private constructor(
        private val window: Window,
        requestedScale: Int,
    ) : AutoCloseable {
        private val previousScale = window.guiScale
        private var isClosed = false

        init {
            window.guiScale = requestedScale
        }

        override fun close() {
            check(!isClosed) { "GUI scale override closed twice" }
            isClosed = true
            window.guiScale = previousScale
        }

        internal companion object {
            fun create(window: Window, requestedScale: Int) = WindowScaleOverride(window, requestedScale)
        }
    }

    companion object {
        private val minecraft = Minecraft.getInstance()
        private var pendingRenderBatch: RenderBatch? = null
        private var overlaysUseNormalCoordinates = false

        @JvmStatic
        fun usesSeparateInventoryScale(screen: Screen?): Boolean =
            config().separateInventoryGuiScale && supportsInventoryScale(screen)

        @JvmStatic
        fun usesSeparateTooltipScale(screen: Screen?): Boolean =
            config().separateTooltipGuiScale && supportsInventoryScale(screen)

        @JvmStatic
        fun resolve(screen: Screen?, window: Window): ResolvedScales {
            val normal = resolve(window, minecraft.options.guiScale().get())
            val tooltip = resolve(window, config().settings.tooltipGuiScale)
            val inventory = minOf(
                resolve(window, config().settings.inventoryGuiScale),
                inventoryScaleLimit(screen),
            )
            return ResolvedScales(normal, inventory.coerceAtLeast(1), tooltip)
        }

        @JvmStatic
        fun useInventoryScale(screen: Screen?, window: Window): WindowScaleOverride =
            WindowScaleOverride.create(window, resolve(screen, window).inventory())

        @JvmStatic
        fun useTooltipScale(screen: Screen?, window: Window): WindowScaleOverride =
            WindowScaleOverride.create(window, resolve(screen, window).tooltip())

        @JvmStatic
        fun convertCoordinate(coordinate: Int, sourceScale: Int, targetScale: Int): Int =
            Math.round(coordinate * sourceScale.coerceAtLeast(1) / targetScale.coerceAtLeast(1).toFloat())

        @JvmStatic
        fun updateScreenDimensions(screen: Screen, window: Window) {
            val state = screen as ScaledScreenState
            val width = window.guiScaledWidth
            val height = window.guiScaledHeight
            if (!state.skysoftMatchesScaleDimensions(width, height)) {
                screen.resize(width, height)
                state.skysoftRememberScaleDimensions(width, height)
            }
        }

        @JvmStatic
        fun restoreScreenDimensions(screen: Screen, window: Window) {
            val state = screen as ScaledScreenState
            if (!state.skysoftHasScaleDimensions()) return
            val width = window.guiScaledWidth
            val height = window.guiScaledHeight
            if (!state.skysoftMatchesScaleDimensions(width, height)) screen.resize(width, height)
            state.skysoftForgetScaleDimensions()
        }

        @JvmStatic
        fun submitRenderBatch(inventory: GuiRenderState, overlays: GuiRenderState) {
            check(pendingRenderBatch == null) { "Inventory GUI render batch was not consumed" }
            pendingRenderBatch = RenderBatch(inventory, overlays)
        }

        @JvmStatic
        fun takeRenderBatch(): RenderBatch? = pendingRenderBatch.also { pendingRenderBatch = null }

        @JvmStatic
        fun areOverlaysUsingNormalCoordinates(): Boolean = overlaysUseNormalCoordinates

        @JvmStatic
        fun setOverlaysUseNormalCoordinates(value: Boolean) {
            overlaysUseNormalCoordinates = value
        }

        private fun resolve(window: Window, configuredScale: Int): Int =
            window.calculateScale(configuredScale.coerceAtLeast(0), minecraft.isEnforceUnicode).coerceAtLeast(1)

        private fun supportsInventoryScale(screen: Screen?): Boolean =
            screen is AbstractContainerScreen<*> || screen is InventoryScaledScreen && screen.usesInventoryScale()

        private fun inventoryScaleLimit(screen: Screen?): Int =
            if (screen is InventoryScaledScreen && screen.usesInventoryScale()) {
                screen.inventoryScaleLimit().coerceAtLeast(1)
            } else {
                Int.MAX_VALUE
            }

        private fun config(): InventoryScreenConfig = SkysoftConfigGui.config().gui.inventoryScreen
    }
}
